File: app/src/main/java/com/example/myphotoapp/DetailActivity.java
package com.example.myphotoapp;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcelable;
import android.provider.MediaStore;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.example.myphotoapp.DB.DbOpenHelper;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class DetailActivity extends AppCompatActivity implements View.OnClickListener, ActivityDataListener {
    private static String TAG = DetailActivity.class.getSimpleName();
    private static final int PICK_FROM_ALBUM = 0;
    private static final int DetailUserData = 001;
    EditText et1, et2;
    Button btn_confirm, btn_phview;
    ImageView img_view;
    SQLiteDatabase db;
    DbOpenHelper dbopen;
    private Context mContext;
    private String title, content;
    private Uri ImageUri;
    private Bitmap mBitmap;
    private Drawable dImage;
    private MainActivity mMainActivity;
    private ArrayList<User> mUserlist;
    private DataManager dataManager = DataManager.getInstance();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.edit_detail);
        mContext = getApplicationContext();
        dbopen = new DbOpenHelper(mContext);
        dbopen.open();

        initButton();
        mMainActivity = MainActivity.getInstance();
    }

    public void initButton() {
        et1 = findViewById(R.id.etview1);
        et2 = findViewById(R.id.etview2);
        btn_phview = findViewById(R.id.btn_phview);
        btn_confirm = findViewById(R.id.btn_confirm);
        img_view = findViewById(R.id.img_view);

        btn_confirm.setOnClickListener(this);
        btn_phview.setOnClickListener(this);
        et1.setOnClickListener(this);

        et1.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                Log.d(TAG, editable.toString());
                title = editable.toString();
            }
        });
        et2.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                Log.d(TAG, editable.toString());
                content = editable.toString();
            }
        });

    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.etview1:
                Log.d(TAG, "etview1");
                break;

            case R.id.etview2:
                Log.d(TAG, "etview2");
                break;
            case R.id.btn_phview:
                takeAlbum();
                break;
            case R.id.btn_confirm:
                Log.d(TAG, "btn_confirm");
                mUserlist= new ArrayList<>();
                if (title != null && content != null) {
                    dbopen.insert(title, content, getByteArrayFromDrawable(dImage));

//                    Intent inserdb = new Intent(this, MainActivity.class);
//                    mUserlist = dbopen.read();
//                    inserdb.putExtra("userlist", mUserlist);
//                    startActivityForResult(inserdb,DetailUserData);

                    dbopen.close();
//                    ((MainActivity)getParent()).cycleList(getApplicationContext());
                    MainActivity.instance.cycleList(mContext);
//                    MainActivity.instance=null;

                } else {
                    Toast.makeText(this, "데이터가 비어있습니다.", Toast.LENGTH_LONG).show();
                }

//                MainActivity.getInstance().recyclerAdapter.notifyDataSetChanged();
                break;
            case R.id.img_view:
                Log.d(TAG, "img_view");

                break;
            default:
                Log.d(TAG, "default");
                break;

        }
    }

    public void takeAlbum() {
        Log.d(TAG, "takeAlbum");
        Intent intent = new Intent(Intent.ACTION_PICK);
        intent.setType(MediaStore.Images.Media.CONTENT_TYPE);
        startActivityForResult(intent, PICK_FROM_ALBUM);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode != RESULT_OK)
            return;

        switch (requestCode) {
            case PICK_FROM_ALBUM:
                Log.d(TAG, "PICK_FROM_ALBUM");

                Uri extras = data.getData();
                Log.d(TAG, extras.toString());


                if (extras != null) {
                    try {
                        Bitmap bitmap = MediaStore.Images.Media.getBitmap(this.getContentResolver(), extras);
                        img_view.setImageBitmap(bitmap);
                        dImage = new BitmapDrawable(bitmap);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }


                }


                break;
            default:
                break;

        }
    }

    public byte[] getByteArrayFromDrawable(Drawable d) {
        byte[] data = null;
        if (d != null) {
            Bitmap bitmap = ((BitmapDrawable) d).getBitmap();
            ByteArrayOutputStream stream = new ByteArrayOutputStream();
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream); //png로하면 배경색이 투명, Jpg로하면 배경색이 검정
            data = stream.toByteArray(); // blob
            return data;
        } else {
            return data;
        }

    }

    @Override
    public void onActivityMessageReceived(int actionCode, Bundle data) {

    }
}

File: app/src/main/java/com/example/myphotoapp/DetailActivity.kt
package com.example.myphotoapp

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.database.sqlite.SQLiteDatabase
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity

import com.example.myphotoapp.DB.DB.DbOpenHelper
import com.example.myphotoapp.DB.DB.User
import com.example.myphotoapp.Logger.Logf
import kotlinx.android.synthetic.main.edit_detail.*

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.ArrayList

class DetailActivity : AppCompatActivity(), View.OnClickListener {


    internal var db: SQLiteDatabase? = null
    internal var dbopen: DbOpenHelper?= null
    private var mContext: Context? = null
    private var title: String? = null
    private var content: String? = null
    private val ImageUri: Uri? = null
    private val mBitmap: Bitmap? = null
    private var dImage: Drawable? = null
    private var mMainActivity: MainActivity? = null
    private var mUserlist: ArrayList<User>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.edit_detail)
        mContext = applicationContext
        dbopen = DbOpenHelper(mContext)
        dbopen!!.open()

        initButton()
        mMainActivity = MainActivity.getInstance()
    }

    fun initButton() {

        btn_confirm.setOnClickListener(this)
        btn_phview.setOnClickListener(this)
        etview1.setOnClickListener(this)

        etview1.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun afterTextChanged(editable: Editable) {
                Logf.v(TAG, editable.toString())
                title = editable.toString()
            }
        })
        etview2.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun afterTextChanged(editable: Editable) {
                Logf.v(TAG, editable.toString())
                content = editable.toString()
            }
        })

    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.etview1 -> Logf.v(TAG, "etview1")

            R.id.etview2 -> Logf.v(TAG, "etview2")
            R.id.btn_phview -> takeAlbum()
            R.id.btn_confirm -> {
                Logf.v(TAG, "btn_confirm")
                mUserlist = ArrayList()
                if (title != null && content != null) {
                    if (dImage == null) {
                        dbopen!!.insert(title, content, null)
                    } else {
                        dbopen!!.insert(title, content, getByteArrayFromDrawable(dImage))
                    }


                    dbopen!!.close()
                    MainActivity.getInstance().cycleList(mContext!!)

                } else {
                    Toast.makeText(this, "데이터가 비어있습니다.", Toast.LENGTH_LONG).show()
                }
            }
            R.id.img_view -> Logf.v(TAG, "img_view")
            else -> Logf.v(TAG, "default")
        }
    }

    fun takeAlbum() {
        Logf.v(TAG, "takeAlbum")
        val intent = Intent(Intent.ACTION_PICK)
        intent.type = MediaStore.Images.Media.CONTENT_TYPE
        startActivityForResult(intent, PICK_FROM_ALBUM)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (resultCode != Activity.RESULT_OK)
            return

        when (requestCode) {
            PICK_FROM_ALBUM -> {
                Logf.v(TAG, "PICK_FROM_ALBUM")

                val extras = data!!.data
                Logf.v(TAG, extras!!.toString())


                if (extras != null) {
                    try {
                        val bitmap = MediaStore.Images.Media.getBitmap(this.contentResolver, extras)
                        img_view.setImageBitmap(bitmap)
                        dImage = BitmapDrawable(bitmap)
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }


                }
            }
            else -> {
            }
        }
    }

    fun getByteArrayFromDrawable(d: Drawable?): ByteArray? {
        var data: ByteArray? = null
        if (d != null) {
            val bitmap = (d as BitmapDrawable).bitmap
            val stream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream) //png로하면 배경색이 투명, Jpg로하면 배경색이 검정
            data = stream.toByteArray() // blob
            return data
        } else {
            return data
        }

    }

    //    private byte[] getBlob(String image) {
    //        ByteArrayBuffer baf = new ByteArrayBuffer(500);
    //        try {
    //            String FILE_PATH1 = "sdcard/"
    //            File file = new File(FILE_PATH1, image + ".jpg");
    //            InputStream is = new FileInputStream(file);
    //            BufferedInputStream bis = new BufferedInputStream(is);
    //            int current = 0;
    //            while ((current = bis.read()) != -1) {
    //                baf.append((byte) current);
    //            }
    //            return baf.toByteArray();
    //        } catch (Exception e) {
    //            Log.d("ImageManager", "Error: " + e.toString());
    //        }
    //        return baf.toByteArray();
    //    }


    companion object {
        private val TAG = DetailActivity::class.java.simpleName
        private val PICK_FROM_ALBUM = 0
        private val DetailUserData = 1
    }
}
